File: backend/app/services/member_service.py
from app.core.db import supabase
from app.schemas.member import MemberCreate, MemberUpdate


def get_all_members():
    response = supabase.table("members").select("*").execute()
    return response.data


def get_member_by_id(member_id: str):
    response = supabase.table("members").select("*").eq("id", member_id).single().execute()
    return response.data


def create_member(payload: MemberCreate):
    response = supabase.table("members").insert(payload.model_dump()).execute()
    return response.data[0]

def update_member(member_id: str, payload: MemberUpdate):
    response = supabase.table("members").update(payload.model_dump(exclude_unset=True)).eq("id", member_id).execute()
    return response.data[0]

File: backend/src/services/memberService.ts
import { DbClient } from "../core/dbClient";
import { MemberCreate, MemberUpdate } from "../schemas/member";
import { NotFoundError } from "../core/errors";

export function createMemberService(db: DbClient) {
  return {
    async getAllMembers() {
      const { data, error } = await db.from("members").select("*");
      if (error) throw error;
      return data;
    },

    async getMemberById(memberId: string) {
      const { data, error } = await db.from("members").select("*").eq("id", memberId).single();
      if (error || !data) throw new NotFoundError("Member not found");
      return data;
    },

    async createMember(payload: MemberCreate) {
      const { data, error } = await db.from("members").insert(payload).select().single();
      if (error) throw error;
      return data;
    },

    async updateMember(memberId: string, patch: Partial<MemberUpdate>) {
      const { data, error } = await db
        .from("members")
        .update(patch)
        .eq("id", memberId)
        .select()
        .single();
      if (error || !data) throw new NotFoundError("Member not found");
      return data;
    },

    async deleteMember(memberId: string) {
      const { data, error } = await db
        .from("members")
        .delete()
        .eq("id", memberId)
        .select()
        .single();
      if (error || !data) throw new NotFoundError("Member not found");
      return data;
    },
  };
}

export type MemberService = ReturnType<typeof createMemberService>;
